src/main/java/net/neoforged/neoforge/transfer/transaction/snapshots/GroupedSnapshotJournal.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.transaction.snapshots;
+
+import net.neoforged.neoforge.transfer.transaction.SnapshotJournal;
+import net.neoforged.neoforge.transfer.transaction.TransactionContext;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A snapshot journal with a runnable callback that plays only on a successful commit or on reversion. This is intended to be used when
+ * you may have multiple journals bound to individual indices, but want to act only once on failure/success at the end of the transaction chain.
+ * <p>
+ * A typical example is when you have an inventory of say 100 different indices, and each index has its own journal. This can be set to
+ * update with them, (like {@link IndexedIntSnapshot} does in {@link IndexedIntSnapshot#updateSnapshots(TransactionContext)}) so that any time a new snapshot is taken,
+ * this will update at the same time. Since we only store one snapshot per depth in a transaction, this will only have 1 regardless of which and how many indices were updated in that chain.
+ * <p>
+ * When the transaction is committed, {@link #commitCallback} will run; and when the transaction is reverted {@link #revertCallback} will instead. Only one per transaction chain will be called and only once.
+ */
+public final class GroupedSnapshotJournal extends SnapshotJournal<GroupedSnapshotJournal.IgnoredValue> {
+    @Nullable
+    private final Runnable commitCallback;
+    @Nullable
+    private final Runnable revertCallback;
```

Comment:
This should probably use standard transaction terminology and be named "rollback".

Furthermore, is the "callback" suffix necessary? It might be clearer if these are named "onCommit" or "onRollback"/"onRevert".